drivers/io-expander/demux_spi/demux_spi.h

```diff
@@ -0,0 +1,80 @@
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#ifndef SRC_DEMUX_SPI_H_
+#define SRC_DEMUX_SPI_H_
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include "spi.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+#define CS_OFFSET 0x80
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/**
+ * @struct spi_desc
+ * @brief Structure initialization with the platform specific SPI functions
+ */
+extern const struct spi_platform_ops demux_spi_platform_ops;
+
```

Comment:
I would prefer an other approach to not create a new spi_desc for each spi device is connected to the dmux.
Something like this:
```suggestion
//dmux_spi.h
struct dmux_init_param {
	enum {
		INIT_DEMUX,
		ATTACH_SPI_DEVICE
	} operation;
	union {
		void *extra;
		struct spi_desc *dmux;
	}	ref;
}

//main.c
		struct spi_desc *dmux;
		struct xil_spi_init_param xil_spi_param = {
#ifdef PLATFORM_MB
		.type = SPI_PL,
#else
		.type = SPI_PS,
#endif
		.device_id = SPI_DEVICE_ID
	};
	struct dmux_init_param dmux_param = {
		.operation = INIT_DEMUX,
		.ref.extra = xil_spi_param
	}
	struct spi_init_param demux_spi_param = {
		.max_speed_hz = 2000000u,
		.chip_select = 0,
		.mode = SPI_MODE_0,
		.platform_ops = &xil_platform_ops
		.extra = 
	};

	spi_init(&dmux, &demux_spi_param);

	dmux_param.operation = ATTACH_SPI_DEVICE;
	dmux_param.ref.dmux = dmux;


	struct spi_init_param ad9250_0_spi_param = {
		.max_speed_hz = 2000000u,
		.chip_select = 0,
		.mode = SPI_MODE_0,
		.platform_ops = &demux_spi_platform_ops
		.extra = dmux_param
	};

	struct spi_init_param ad9250_1_spi_param = {
		.max_speed_hz = 2000000u,
		.chip_select = 1,
		.mode = SPI_MODE_0,
		.platform_ops = &demux_spi_platform_ops
		.extra = dmux_param
	};
	
	spi_init(&ad9250_0_spi, &ad9250_0_spi_param);
	spi_init(&ad9250_1_spi, &ad9250_1_spi_param);
	
//dmux_spi.c
struct dmux_desc {
	spi_desc *dev;
	spi_desc *current_dev;
}

dmux_init()
{
	switch(init_param->extra->operation)
	 case INIT_DEMUX:
		*desc = calloc();
		spi_init_param new_param;
		new_param = param;
		new_param.extra = param->extra->extra;
		spi_init(desc->dev, new_param); //here is platform initialization
	 case ATTACH_SPI_DEVICE:
		*desc = param->extra->dmux;
		desc->current_dev = calloc(sizeof(spi_desc));
		*desc->current_dev.chip_select = param.chip_select;
}

dmux_write_and_read(){
	cs = CS_OFFSET | current_dev->chip_select;
	spi_write_and_read(desc->dev, ...);
}

```

Reply:
The solution proposed by this pull-request seems to be generic enough to accommodate any SPI configuration.
With your approach, I don't think that the following example will work:
Platform_SPI (CS_0) -> MUX_SPI_0 (CS_0) -> MUX_SPI_1 (CS_0) -> ADC_0
                                                                                                 (CS_1) -> ADC_1
                                                                                                  ...
                                                                                                 (CS_n) -> ADC_n
                                                            (CS_1) -> DAC_0
                                                            (CS_2) -> DAC_1
                    (CS_1) -> DAC_2

Reply:
![image](https://user-images.githubusercontent.com/16192750/80979593-39f2a880-8e30-11ea-9f08-cfd77707a4de.png)

This is how initialization for your example should look like with this approach.
Maybe in the sketch I didn't think about all the implementation details, but I only wanted to point the bigger picture.
It is just a suggestion that makes more sens to me instead of how it is implemented now:

![image](https://user-images.githubusercontent.com/16192750/80980592-7bd01e80-8e31-11ea-90b1-2a12dd056add.png)